src/components/NewClusterModal.js

```diff
@@ -42,386 +43,717 @@ const imageValidationRegexp = /^[A-Za-z0-9]+[\w./-]+(?::\w[\w.-]+)?(?:@[\w+.-]+:
-      currentClusterConfig.numberOfWorkerLocalSSDs !== userSelectedConfig.numberOfWorkerLocalSSDs
+    this.sendCloudEnvironmentMetrics()
 
-    const hasWorkers = currentClusterConfig.numberOfWorkers >= 2 || currentClusterConfig.numberOfPreemptibleWorkers >= 2
-    const hasWorkersResourceChanged = hasWorkers && hasUnUpdateableResourceChanged
+    if (shouldDeleteRuntime) {
+      await Ajax().Clusters.cluster(namespace, currentClusterDetails.runtimeName).delete(this.hasAttachedDisk() && shouldDeletePersistentDisk)
+    }
+    if (shouldDeletePersistentDisk && !this.hasAttachedDisk()) {
+      await Ajax().Disks.disk(namespace, currentPersistentDiskDetails.name).delete()
+    }
+    if (shouldUpdatePersistentDisk) {
+      await Ajax().Disks.disk(namespace, currentPersistentDiskDetails.name).update(newPersistentDisk.size)
+    }
+    if (shouldUpdateRuntime) {
+      await Ajax().Clusters.cluster(namespace, currentClusterDetails.runtimeName).update({ runtimeConfig })
+    }
+    if (shouldCreateRuntime) {
+      await Ajax().Clusters.cluster(namespace, Utils.generateClusterName()).create({
+        runtimeConfig,
+        toolDockerImage: newRuntime.toolDockerImage,
+        labels: { saturnIsProjectSpecific: `${selectedLeoImage === PROJECT_SPECIFIC_MODE}` },
+        ...(newRuntime.jupyterUserScriptUri ? { jupyterUserScriptUri: newRuntime.jupyterUserScriptUri } : {})
+      })
+    }
 
-    const hasDiskSizeDecreased = currentClusterConfig.masterDiskSize > userSelectedConfig.masterDiskSize
+    onSuccess()
+  })
 
-    const hasCloudServiceChanged = currentClusterConfig.cloudService !== userSelectedConfig.cloudService
+  getNewEnvironmentConfig() {
+    const {
+      deleteDiskSelected, selectedPersistentDiskSize, viewMode, masterMachineType,
+      masterDiskSize, sparkMode, numberOfWorkers, numberOfPreemptibleWorkers, workerMachineType,
+      workerDiskSize, jupyterUserScriptUri, selectedLeoImage, customEnvImage
+    } = this.state
+    const { persistentDisk: oldPersistentDisk, runtime: oldRuntime } = this.getOldEnvironmentConfig()
+    const cloudService = sparkMode ? cloudServices.DATAPROC : cloudServices.GCE
+    const newNumberOfWorkers = sparkMode === 'cluster' ? numberOfWorkers : 0
+    return {
+      runtime: Utils.cond(
+        [(viewMode !== 'deleteEnvironmentOptions'), () => {
+          return {
+            cloudService,
+            ...(cloudService === cloudServices.GCE ? {
+              machineType: masterMachineType,
+              ...(this.shouldUsePersistentDisk() ? {
+                persistentDiskAttached: true
+              } : {
+                diskSize: masterDiskSize
+              })
+            } : {
+              masterMachineType,
+              masterDiskSize,
+              numberOfWorkers: newNumberOfWorkers,
+              ...(newNumberOfWorkers && {
+                numberOfPreemptibleWorkers,
+                workerMachineType,
+                workerDiskSize
+              })
+            }),
```

Comment:
Since there is a bunch of common stuff and a big block of stuff that is different based on the type of runtime, I think it might read well to extract the big block of conditional stuff. Then this object construction is a straightforward 4 lines with only inline conditionals.
```suggestion
    const runtimeDetails = {
      ...(cloudService === cloudServices.GCE ? {
        machineType: masterMachineType,
        ...(this.shouldUsePersistentDisk() ? {
          persistentDiskAttached: true
        } : {
          diskSize: masterDiskSize
        })
      } : {
        masterMachineType,
        masterDiskSize,
        numberOfWorkers: newNumberOfWorkers,
        ...(newNumberOfWorkers && {
          numberOfPreemptibleWorkers,
          workerMachineType,
          workerDiskSize
        })
      })
    }
    return {
      runtime: Utils.cond(
        [(viewMode !== 'deleteEnvironmentOptions'), () => {
          return {
            cloudService,
            ...runtimeDetails,
```
Does that work for you @s-rubenstein ?

Reply:
I took Sky's suggestion here. I think keeping the data inline makes it easier to see the shape of this structure.